[Form Control Refresh] Background color for native text inputs does not change when autofilled

lilyspiniolas opened the pull request:
#### 29e6f4d3eec462798020e2f1a96b071455b96133
<pre>
[Form Control Refresh] Background color for native text inputs does not change when autofilled
<a href="https://bugs.webkit.org/show_bug.cgi?id=297642">https://bugs.webkit.org/show_bug.cgi?id=297642</a>
<a href="https://rdar.apple.com/158720549">rdar://158720549</a>

Reviewed by Aditya Keerthi.

During textarea/textfield painting, use the background color of the renderer&apos;s style so that
the autofill background color applies.

Added tests fast/forms/form-control-refresh/autofilled-and-obscured-text-input.html and
fast/forms/form-control-refresh/autofilled-text-input.html to ensure the correct background
and text colors are used for native text inputs when autofilled, and when autofilled + obscured.

* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-dark-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-light-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-dark-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-light-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/resources/autofill-test.js: Added.
(test.hideElementBorders):
(test):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::paintTextAreaOrTextField):

Canonical link: <a href="https://commits.webkit.org/299088@main">https://commits.webkit.org/299088@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/402c057f7cbbe329e1274dd252f34d7ece3b643a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117780 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123916 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69800 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/73c18905-4959-49f2-b1e8-691490089d88) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119658 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46040 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89386 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/58576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e0768d1f-eebf-41ef-8d9a-74bbe26ca354) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120732 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30372 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105607 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69879 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c85429a1-978b-4525-9150-dc1bd92e873a) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29436 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23722 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67579 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127007 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44683 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33638 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98044 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45041 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101832 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97832 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24892 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43218 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21187 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41049 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44555 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50229 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44013 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47360 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45704 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7aa35e2ef8ad0700c3d17b2f051f70587e6bc1c3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49642

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -2170,7 +2170,14 @@ static bool paintTextAreaOrTextField(const RenderObject& box, const PaintInfo& p
 #endif
 
     const auto styleColorOptions = box.styleColorOptions();
-    auto backgroundColor = RenderTheme::singleton().systemColor(CSSValueCanvas, styleColorOptions);
```
I wonder if we can just read the background color from the style? Since `background-color: Canvas;` is specified by default in the UA stylesheet.

lilyspiniolas replied:
Since this is theme code, in my opinion it should generally be agnostic from the UA stylesheet, but if you feel strongly I'm not opposed to making this change

pxlcoder replied:
I don't think there's any issue with reading from style. We're not specifically reading from the UA stylesheet, just from `box.style()` which we already do in many places.

If it works, there are multiple benefits to doing so here:
- Improved maintainability
  - If someone wanted to modify the autofill color, they would have to update two places
  - Similarly, if someone wanted to introduce a new autofill pseudo-class
- Reduced duplication

lilyspiniolas replied:
Those are good points, I'll go ahead and update it.

pxlcoder commented:
I'd also update the commit/bug title to say *native* text inputs.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3645b2bae791bff1293358d0806959a17d730ba8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49642

pxlcoder commented on LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-expected.html:
```diff
@@ -0,0 +1,26 @@
+<head>
+    <script>
+        function setBackground() {
+            var tf = document.getElementById('devolved');
```
Nit: Here and elsewhere, we should use full words: https://webkit.org/code-style-guidelines/#names-full-words

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -2170,7 +2170,7 @@ static bool paintTextAreaOrTextField(const RenderObject& box, const PaintInfo& p
 
     const auto styleColorOptions = box.styleColorOptions();
-    auto backgroundColor = RenderTheme::singleton().systemColor(CSSValueCanvas, styleColorOptions);
+    auto backgroundColor = style->backgroundColor().resolveColor(style->color());
```
```suggestion
    auto backgroundColor = style->visitedDependentColor(CSSPropertyBackgroundColor);
```

pxlcoder reviewed (commented):
Please make sure to (manually) test dark mode too. For both autofilled content, and the default native appearance.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5c3051f8fa3d32d66d174a20b5466e2a3e1141a6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49642

lilyspiniolas committed:
[Form Control Refresh] Background color for native text inputs does not change when autofilled
https://bugs.webkit.org/show_bug.cgi?id=297642
rdar://158720549

Reviewed by Aditya Keerthi.

During textarea/textfield painting, use the background color of the renderer's style so that
the autofill background color applies.

Added tests fast/forms/form-control-refresh/autofilled-and-obscured-text-input.html and
fast/forms/form-control-refresh/autofilled-text-input.html to ensure the correct background
and text colors are used for native text inputs when autofilled, and when autofilled + obscured.

* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-dark-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-light-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-and-obscured-text-input-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-dark-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-light-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/autofilled-text-input-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/resources/autofill-test.js: Added.
(test.hideElementBorders):
(test):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::paintTextAreaOrTextField):

Canonical link: https://commits.webkit.org/299088@main
